File: ios/Khrysalis Template/com/lightningkite/khrysalistemplate/vg/PreviewVG.shared.swift
//Package: com.lightningkite.khrysalistemplate.vg
//Converted using Khrysalis2

import Foundation
import Khrysalis
import RxSwift
import RxRelay
import KhrysalisMaps



public class PreviewVG: ViewGenerator {
    
    
    override public var title: String {
        get {
            return "Preview"
        }
    }
    
    public class XmlPreview {
        
        public var name: String
        public var make:  (ViewDependency) -> View
        
        
        public init(name: String, make: @escaping (ViewDependency) -> View) {
            self.name = name
            self.make = make
        }
        convenience public init(_ name: String, _ make: @escaping (ViewDependency) -> View) {
            self.init(name: name, make: make)
        }
    }
    public var previews: Array<XmlPreview>
    public var previewIndex: StandardObservableProperty<Int32>
    
    override public func generate(dependency: ViewDependency) -> View {
        var xml = PreviewXml()
        var view = xml.setup(dependency)
        xml.pager.bind(previews, previewIndex) { (it) in 
            it.make(dependency)
        }
        previewIndex.addAndRunWeak(self, xml.viewName) { (self, view, it) in 
            view.text = self.previews[ it ].name
        }
        return view
    }
    override public func generate(_ dependency: ViewDependency) -> View {
        return generate(dependency: dependency)
    }
    
    override public init() {
        let previews: Array<XmlPreview> = [XmlPreview("ControlsDemoXml") { (it) in 
            ControlsDemoXml().setup(it)
        }, XmlPreview("DateRangeDemoXml") { (it) in 
            DateRangeDemoXml().setup(it)
        }, XmlPreview("ExampleContentXml") { (it) in 
            ExampleContentXml().setup(it)
        }, XmlPreview("LoginDemoXml") { (it) in 
            LoginDemoXml().setup(it)
        }, XmlPreview("MainXml") { (it) in 
            MainXml().setup(it)
        }, XmlPreview("SegmentedControlDemoXml") { (it) in 
            SegmentedControlDemoXml().setup(it)
        }, XmlPreview("SelectDemoXml") { (it) in 
            SelectDemoXml().setup(it)
        }, XmlPreview("SliderDemoXml") { (it) in 
            SliderDemoXml().setup(it)
        }, XmlPreview("ViewPagerDemoXml") { (it) in 
            ViewPagerDemoXml().setup(it)
        }]
        self.previews = previews
        let previewIndex: StandardObservableProperty<Int32> = StandardObservableProperty(0)
        self.previewIndex = previewIndex
        super.init()
    }
}
 

File: android/src/main/java/com/lightningkite/khrysalistemplate/vg/PreviewVG.shared.kt
package com.lightningkite.khrysalistemplate.vg

import android.view.View
import com.lightningkite.khrysalis.escaping
import com.lightningkite.khrysalis.observables.binding.bind
import com.lightningkite.khrysalis.observables.StandardObservableProperty
import com.lightningkite.khrysalis.observables.addAndRunWeak
import com.lightningkite.khrysalis.observables.subscribeBy
import com.lightningkite.khrysalis.rx.removed
import com.lightningkite.khrysalis.rx.until
import com.lightningkite.khrysalis.views.ViewDependency
import com.lightningkite.khrysalis.views.ViewGenerator
import com.lightningkite.khrysalistemplate.layouts.*

class PreviewVG : ViewGenerator() {
    override val title: String
        get() = "Preview"

    class XmlPreview(val name: String, val make: @escaping() (ViewDependency) -> View)

    val previews: List<XmlPreview> = listOf(
        XmlPreview("ControlsDemoXml") { it -> ControlsDemoXml().setup(it) },
        XmlPreview("DateRangeDemoXml") { it -> DateRangeDemoXml().setup(it) },
        XmlPreview("ExampleContentXml") { it -> ExampleContentXml().setup(it) },
        XmlPreview("LoginDemoXml") { it -> LoginDemoXml().setup(it) },
        XmlPreview("MainXml") { it -> MainXml().setup(it) },
        XmlPreview("SegmentedControlDemoXml") { it -> SegmentedControlDemoXml().setup(it) },
        XmlPreview("SelectDemoXml") { it -> SelectDemoXml().setup(it) },
        XmlPreview("SliderDemoXml") { it -> SliderDemoXml().setup(it) },
        XmlPreview("ViewPagerDemoXml") { it -> ViewPagerDemoXml().setup(it) }
    )

    val previewIndex: StandardObservableProperty<Int> = StandardObservableProperty(0)

    override fun generate(dependency: ViewDependency): View {
        val xml = PreviewXml()
        val view = xml.setup(dependency)
        xml.pager.bind(previews, previewIndex) { it -> it.make(dependency) }
        previewIndex.subscribeBy { it ->
            xml.viewName.text = this.previews[it].name
        }.until(view.removed)
        return view
    }

}
